src/TraceEvent/EventPipe/EventCache.cs

```diff
@@ -0,0 +1,282 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Runtime.InteropServices;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Microsoft.Diagnostics.Tracing.EventPipe
+{
+    unsafe delegate void ParseBufferItemFunction(ref EventPipeEventHeader header);
+
+    internal class EventCache
+    {
+        public event ParseBufferItemFunction OnEvent;
+        public event Action<int> OnEventsDropped;
+
+        public unsafe void ProcessEventBlock(byte[] eventBlockData)
+        {
+            // parse the header
+            if(eventBlockData.Length < 20)
+            {
+                Debug.Assert(false, "Expected EventBlock of at least 20 bytes");
+                return;
+            }
+            ushort headerSize = BitConverter.ToUInt16(eventBlockData, 0);
+            if(headerSize < 20 || headerSize > eventBlockData.Length)
+            {
+                Debug.Assert(false, "Invalid EventBlock header size");
+                return;
+            }
+            ushort flags = BitConverter.ToUInt16(eventBlockData, 2);
+            bool useHeaderCompression = (flags & (ushort)EventBlockFlags.HeaderCompression) != 0;
+
+            // parse the events
+            PinnedBuffer buffer = new PinnedBuffer(eventBlockData);
```

Comment:
I'm wondering if you need to pin using a handle or if you can pin using the fixed keyword.  I see one place where the buffer gets handed out, but not sure if it must be pinned for that operation.  Unlikely to hurt for smaller traces, but for very large traces, I could see this having an impact.

Reply:
We could pin/un-pin the event buffer + stack buffer on every dispatch, but I didn't expect that to perform better than pinning the buffers once up-front. If performance profiling revealed it was an issue and fixed performed better I certainly have no qualms switching, its just a matter of doing investigative work in the first place. The perf criteria for file load speed so far has been 'manual testing doesn't show it is clearly slower than other trace formats on modest input sizes.' So nothing overly rigorous : )

Reply:
From a separate angle I'd be curious to hear from you how hard it might be to get rid of the pinning + raw pointer passing in this code path entirely? Certainly we can harden it and still use pointers, but it is far more error prone vs using byte[] + index or Span where the runtime will range check everything automatically. The part I haven't investigated is that currently the contract for all Sources is that they return an EVENT_RECORD*, and then there are more pointers inside of that structure. The API contract would have to change, as would the upstream callers.

Reply:
I've not looked into it, but I expect that we could Span-ify this code.  I'm not sure how easy it would be to switch to byte[] + index as there are some places where we operate on native buffers and wouldn't want to do the copy.  This just hasn't been Span-ified because it hasn't come to the front of the priority queue.

In general, fixed doesn't have a cost runtime until there is a GC and that code is on the stack, so that's why it's preferrable from a performance POV.  If the creation/destruction of the handle doesn't happen very often, then it's not interesting at all.  I think I'm concerned about this more as a scalability thing than anything else.  I don't see this as blocking, but figured that if it was super-cheap to switch to fixed then it could be worthwhile.  I'm also OK deferring this unless/until we need it.  Out of curiosity, how large are these blocks?

Reply:
Current limit is 100KB on the blocks, but they could be much smaller if the stream is being flushed frequently and a low volume of events are flowing through.